linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/Preprocessor.scala

```diff
@@ -0,0 +1,458 @@
+   *     +--------------------+\            packed encoding version
+   *     |       ...          + +-----> [(itable of A), (itable of G), null, null]
+   *     +--------------------+
+   *  }}}
+   *
+   *  To perform an interface dispatch, we can use bucket IDs and indices to
+   *  locate the appropriate interface table. For instance, suppose we need to
+   *  dispatch for interface G. Knowing that G belongs to bucket 1, we retrieve
+   *  the itable for G from i-th element of the itables.
+   *
+   *  @note
+   *    Why the types in the example are assigned to the buckets like that?
+   *    - bucket0: [A]
+   *      - A is placed alone in the first bucket.
+   *      - It cannot be grouped with any of its subtypes as that would violate
+   *        the "no common subtypes" rule.
+   *    - bucket1: [B, C, G]
+   *      - B, C, and G cannot be in the same bucket with A since they are all
+   *        direct subtypes of A.
+   *      - They are grouped together because they do not share any common subtype.
+   *    - bucket2: [D, F]
+   *      - D cannot be assigned to neither bucket 0 or 1 because it shares the
+   *        same subtype (D itself) with A (in bucket 0) and C (in bucket 1).
+   *      - D and F are grouped together because they do not share any common subtype.
+   *    - bucket3: [E]
+   *     - E shares its subtype with all the other buckets, so it gets assigned
+   *       to a new bucket.
+   *
+   *  @return
+   *    A sequence of all the buckets, where each bucket is a list of the
+   *    interfaces that were assigned to (the index of) that bucket.
+   *
+   *  @see
+   *    The algorithm is based on the "packed encoding" presented in the paper
+   *    "Efficient Type Inclusion Tests"
+   *    [[https://www.researchgate.net/publication/2438441_Efficient_Type_Inclusion_Tests]]
+   */
+  private def computeItableBuckets(allClasses: List[LinkedClass],
+      isInterface: ClassName => Boolean): Seq[List[ClassName]] = {
+
+    val classes = allClasses.filterNot(_.kind.isJSType)
+
+    /* The algorithm separates the type hierarchy into three disjoint subsets:
+     *
+     * - join types: types with multiple parents (direct supertypes) that have
+     *   only single subtyping descendants:
+     *   `join(T) = {x ∈ multis(T) | ∄ y ∈ multis(T) : y <: x}`
+     *   where multis(T) means types with multiple direct supertypes.
+     * - spine types: all ancestors of join types:
+     *   `spine(T) = {x ∈ T | ∃ y ∈ join(T) : x ∈ ancestors(y)}`
+     * - plain types: types that are neither join nor spine types
+     *
+     * The bucket assignment process consists of two parts:
+     *
+     * **1. Assign buckets to spine types**
+     *
+     * Two spine types can share the same bucket only if they do not have any
+     * common join type descendants.
+     *
+     * Visit spine types in reverse topological order because (from leaves to
+     * root) when assigning a spine type to a bucket, the algorithm already has
```

Comment:
```suggestion
     * Visit spine types in reverse topological order (from leaves to root)
     * because when assigning a spine type to a bucket, the algorithm already has
```